src/interaction/grab.rs

```diff
@@ -0,0 +1,139 @@
+use crate::interaction::{
+    InteractionPlayer, MarkedItem, detect_interactables,
+};
+use bevy::prelude::*;
+use bevy::reflect::Reflect;
+
+/// Plugin that sets up grabbing logic for interactable items.
+pub(super) struct GrabPlugin;
+
+impl Plugin for GrabPlugin {
+    fn build(&self, app: &mut App) {
+        app.init_resource::<GrabState>()
+            .add_event::<GrabEvent>()
+            .add_event::<ReleaseEvent>()
+            .add_systems(
+                Update,
+                (
+                    grab_input_system.after(detect_interactables),
+                    handle_grab,
+                    handle_release,
+                    update_snapping,
+                ),
+            );
+        app.register_type::<Grabbable>().register_type::<Occupied>();
+    }
+}
+
+/// Reads the `E` key press and the current `MarkedItem` to send grab or release events without PlayerAction.
+// TODO: Use PlayerAction instead of KeyCode
+fn grab_input_system(
+    keys: Res<ButtonInput<KeyCode>>,
+    marked_q: Query<(Entity, &MarkedItem), With<InteractionPlayer>>,
+    grabbable_q: Query<&Grabbable>,
+    mut grab_writer: EventWriter<GrabEvent>,
+    mut release_writer: EventWriter<ReleaseEvent>,
+    grab_state: Res<GrabState>,
+) {
+    if keys.just_pressed(KeyCode::KeyE) {
+        if grab_state.held.is_some() {
+            release_writer.write(ReleaseEvent);
+        } else if let Ok((_, marked)) = marked_q.single() {
+            if let Some(target) = marked.0 {
+                if grabbable_q.get(target).is_ok() {
+                    // Send grab event
+                    grab_writer.write(GrabEvent(target));
+                }
+            }
+        }
+    }
+}
+
+/// Attaches the grabbed entity to the player and marks the player occupied.
+fn handle_grab(
+    mut commands: Commands,
+    mut grab_state: ResMut<GrabState>,
+    mut events: EventReader<GrabEvent>,
+    player_q: Query<Entity, With<InteractionPlayer>>,
+) {
+    for GrabEvent(entity) in events.read() {
+        if grab_state.held.is_none() {
+            if let Ok(player) = player_q.single() {
+                commands.entity(player).add_child(*entity);
+                commands.entity(player).insert(Occupied);
+                grab_state.held = Some(*entity);
+            }
+        }
+    }
+}
+
+/// Detaches the held entity and places it in front of the player.
+fn handle_release(
+    mut commands: Commands,
+    mut grab_state: ResMut<GrabState>,
+    mut events: EventReader<ReleaseEvent>,
+    player_q: Query<Entity, With<InteractionPlayer>>,
+    player_tf_q: Query<&GlobalTransform, With<InteractionPlayer>>,
+    mut tf_q: Query<&mut Transform>,
+) {
+    // Release distance from player
+    const RELEASE_DISTANCE: f32 = 2.0;
+    for _ in events.read() {
+        if let Some(entity) = grab_state.held.take() {
+            if let Ok(player) = player_q.single() {
```

Comment:
This is abit dangerous as we might want to support multiplayer, and this would break.